File: app/src/androidTest/java/app/tez/daggertuto/MainActivityAndroidTest.java
package app.tez.daggertuto;




import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import androidx.test.platform.app.InstrumentationRegistry;
import androidx.test.rule.ActivityTestRule;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import app.tez.daggertuto.di.AndroidTestComponent;
import app.tez.daggertuto.di.AppModule;
import app.tez.daggertuto.di.DaggerAndroidTestComponent;
import app.tez.daggertuto.di.StorageModule;
import app.tez.daggertuto.di.UtilsTestModule;
import app.tez.daggertuto.utils.SharedPrefsUtils;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.action.ViewActions.typeText;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static junit.framework.TestCase.assertEquals;

/**
 * Created by Terence Zafindratafa on 30/10/2018
 */
@RunWith(AndroidJUnit4.class)
public class MainActivityAndroidTest {

    @Rule
    public ActivityTestRule<MainActivity> mActivityTestRule = new ActivityTestRule<MainActivity>(MainActivity.class);

    @Before
    public void setup(){
        DaggerTutoApplication app = (DaggerTutoApplication) InstrumentationRegistry.getInstrumentation().getTargetContext().getApplicationContext();

        AndroidTestComponent component = DaggerAndroidTestComponent.builder()
                .appModule(new AppModule(app))
                .storageModule(new StorageModule())
                .utilsTestModule(new UtilsTestModule())
                .build();
        app.tutoComponent = component;
        component.inject(this);
    }

    @Test
    public void isMainActivityOpening(){
        assertEquals(mActivityTestRule.getActivity().getClass().getName(), MainActivity.class.getName());
    }

    @Test
    public void isTheRightDateDisplaying(){
        onView(withId(R.id.activity_main_date)).check(matches(withText("Sat Oct 19 00:00:00 GMT+01:00 1991")));
    }

    @Test
    public void isTheRightEmailSaving(){
        // Before test
        SharedPrefsUtils prefsUtils = mActivityTestRule.getActivity().prefs;
        prefsUtils.saveCurrentEmail("elvis@hotmail.com");

        // Actual test
        onView(withId(R.id.activity_main_editText)).perform(typeText("hello@world.fr"));
        onView(withId(R.id.activity_main_button)).perform(click());
        onView(withId(R.id.activity_main_saved_email)).check(matches(withText("hello@world.fr")));
    }
}

File: app/src/androidTest/java/app/tez/daggertuto/MainActivityAndroidTest.kt
package app.tez.daggertuto



import android.support.test.runner.AndroidJUnit4
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.rule.ActivityTestRule
import app.tez.daggertuto.base.InstTestDaggerTutoApplication
import app.tez.daggertuto.di.DaggerTutoComponent
import app.tez.daggertuto.di.UtilsModule
import app.tez.daggertuto.di.StorageModule
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import java.util.*

/**
 * Created by Terence Zafindratafa on 26/10/2018
 */

private const val MATCHING_DATE = "19/09/1991"

@RunWith(AndroidJUnit4::class)
class MainActivityAndroidTest {

    @Rule
    val activityMainTestRule: ActivityTestRule<MainActivity> = object: ActivityTestRule<MainActivity>(MainActivity::class.java){
        override fun beforeActivityLaunched() {
            val myApp = InstrumentationRegistry.getInstrumentation().targetContext as InstTestDaggerTutoApplication

            val cal = Calendar.getInstance()
            cal.set(Calendar.DAY_OF_MONTH, 19)
            cal.set(Calendar.MONTH, Calendar.SEPTEMBER)
            cal.set(Calendar.YEAR, 1991)

            val component = DaggerTutoComponent.builder()
                    .appModule(myApp)
                    .build()

            myApp.component = component
        }
    }

    /*@Before
    fun setUp() {
        MockitoAnnotations.initMocks(this)
        val myApp = InstrumentationRegistry
                .getInstrumentation()
                .targetContext
                .applicationContext as InstTestDaggerTutoApplication

        val cal = Calendar.getInstance()
        cal.set(Calendar.DAY_OF_MONTH, 19)
        cal.set(Calendar.MONTH, Calendar.SEPTEMBER)
        cal.set(Calendar.YEAR, 1991)

        val component = DaggerTutoComponent.builder()
                .prefModule(StorageModule(myApp))
                .dateModule(UtilsModule(cal))
                .build()

        myApp.component = component
    }*/

    @Test
    fun shouldShowProperDate(){
        onView(withId(R.id.tv)).perform(click())
        onView(withId(R.id.tv)).check(matches(withText(MATCHING_DATE)))
    }


}